framework/resource_cache.h

```diff
@@ -80,82 +68,32 @@ class ResourceCache
-	                                                   const std::vector<ShaderResource> &set_resources);
-
-	GraphicsPipeline &request_graphics_pipeline(PipelineState &pipeline_state);
-
-	ComputePipeline &request_compute_pipeline(PipelineState &pipeline_state);
-
-	DescriptorSet &request_descriptor_set(DescriptorSetLayout &                     descriptor_set_layout,
-	                                      const BindingMap<VkDescriptorBufferInfo> &buffer_infos,
-	                                      const BindingMap<VkDescriptorImageInfo> & image_infos);
-
-	RenderPass &request_render_pass(const std::vector<Attachment> &   attachments,
-	                                const std::vector<LoadStoreInfo> &load_store_infos,
-	                                const std::vector<SubpassInfo> &  subpasses);
-
-	Framebuffer &request_framebuffer(const RenderTarget &render_target,
-	                                 const RenderPass &  render_pass);
-
-	void clear_pipelines();
+	virtual ShaderModule        &request_shader_module(VkShaderStageFlagBits stage, const ShaderSource &glsl_source, const ShaderVariant &shader_variant = {});
+	virtual PipelineLayout      &request_pipeline_layout(const std::vector<ShaderModule *> &shader_modules);
+	virtual DescriptorSetLayout &request_descriptor_set_layout(const uint32_t set_index, const std::vector<ShaderModule *> &shader_modules, const std::vector<ShaderResource> &set_resources);
+	virtual GraphicsPipeline    &request_graphics_pipeline(PipelineState &pipeline_state);
+	virtual ComputePipeline     &request_compute_pipeline(PipelineState &pipeline_state);
+	virtual DescriptorSet       &request_descriptor_set(DescriptorSetLayout &descriptor_set_layout, const BindingMap<VkDescriptorBufferInfo> &buffer_infos, const BindingMap<VkDescriptorImageInfo> &image_infos);
+	virtual RenderPass          &request_render_pass(const std::vector<Attachment> &attachments, const std::vector<LoadStoreInfo> &load_store_infos, const std::vector<SubpassInfo> &subpasses);
+	virtual Framebuffer         &request_framebuffer(const RenderTarget &render_target, const RenderPass &render_pass);
 
 	/// @brief Update those descriptor sets referring to old views
 	/// @param old_views Old image views referred by descriptor sets
 	/// @param new_views New image views to be referred
-	void update_descriptor_sets(const std::vector<core::ImageView> &old_views, const std::vector<core::ImageView> &new_views);
-
-	void clear_framebuffers();
+	virtual void update_descriptor_sets(const std::vector<core::ImageView> &old_views, const std::vector<core::ImageView> &new_views);
 
-	void clear();
+	virtual void clear();
+	virtual void clear_framebuffers();
 
-	const ResourceCacheState &get_internal_state() const;
-
-  private:
+  protected:
 	Device &device;
 
-	ResourceRecord recorder;
-
-	ResourceReplay replayer;
-
-	VkPipelineCache pipeline_cache{VK_NULL_HANDLE};
-
+	mutable std::mutex mutex;
```

Comment:
This mutex is unused?

Reply:
Yup removed. I think it was something i was playing with